refactor(cuda): Refactor the amortized PBS to support k > 1

pdroalves opened the pull request:
### Resolves: https://github.com/zama-ai/concrete-core-internal/issues/448

### Description

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [ ] The draft release description has been updated
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

<!--
### Requires: `<link_your_required_issue_here>`
-->

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

pdroalves commented:
To add support for k>1 in the amortized PBS I had to refactor and fix things all over the crate, including some common methods to the low latency kernel.

pdroalves commented:
@slab-ci gpu_test

agnesLeroy commented on concrete-cuda/cuda/src/bootstrap_amortized.cuh:
```diff
@@ -338,11 +295,11 @@ __host__ void host_bootstrap_amortized(
   }
   checkCudaErrors(cudaGetLastError());
-
+  cudaDeviceSynchronize();
```
I think we should remove this synchronize from here: in the PR where I fixed the asynchronous behaviour I removed them all, and called them from the tests or the Rust integration instead.

pdroalves replied:
Ops, this is a debug artifact. I forgot to remove :sweat_smile:.

agnesLeroy commented on concrete-cuda/cuda/src/crypto/gadget.cuh:
```diff
@@ -11,46 +11,54 @@ private:
-      state[tid + params::degree / 2] += carry_im;
-      res_re -= carry_re << base_log;
-      res_im -= carry_im << base_log;
+    for (int j = 0; j < num_poly; j++) {
```
Here I can see how the bootstrap code is simpler by doing the decomposition on every mask/body element in the same function, but we should probably change the function name to something like `decompose_and_compress_next_glwe` (name to be discussed) to make it clearer in the name that it's not only one value that's being dealt with. Also, this deserves an explanation in a comment for the function.

pdroalves replied:
Something that is not clear to me is how generic this function is (as well as all functions in `functions.cuh`). As I see, this gadget class can be used to decompose any number of polynomials concatenated in an array. These can be a GLWE with any dimension, but also anything else. For instance, I have yet to start the refactoring of the low latency kernel, but I believe one path we could take there is to have different y-blocks handling each polynomial in the mask. In that case, we would need each `decompose_and_compress_next` to decompose a single polynomial and not the entire GLWE. So I'm not very supportive in using "glwe" there unless we have sure we will never use these functions to anything than on GLWEs. What do you think?

Otherwise, I think just adding a comment explaining the class would be enough.

agnesLeroy replied:
Ok agreed, a comment should be enough for us to keep track of what we can do with the function.

agnesLeroy commented on concrete-cuda/cuda/src/polynomial/functions.cuh:
```diff
@@ -62,33 +62,38 @@ __device__ void add_polynomial_inplace_low_lat(T *source, T *dst, int p_id) {
-        uint32_t jj = j - degree;
-        if (tid < degree - jj) {
-          accumulator[tid] = -input[tid + jj];
+  for (int z = 0; z < (glwe_dimension + 1); z++) {
```
I have the same remark here: the function name should be adapted to reflect this change, and a comment should be added to make it clear that it handles all GLWE dimensions, not just one value.

agnesLeroy commented on concrete-cuda/cuda/src/polynomial/functions.cuh:
```diff
@@ -98,28 +103,31 @@ __device__ void divide_by_monomial_negacyclic_inplace(T *accumulator, T *input,
-    if (j < degree) {
-      if (tid < j) {
-        result_acc[tid] = -acc[tid - j + degree] - acc[tid];
+  for (int z = 0; z < (glwe_dimension + 1); z++) {
```
Same here.

agnesLeroy commented on concrete-cuda/cuda/src/polynomial/functions.cuh:
```diff
@@ -129,19 +137,23 @@ multiply_by_monomial_negacyclic_and_sub_polynomial(T *acc, T *result_acc,
+                                                  int level_count,
+                                                  uint32_t glwe_dimension) {
+  constexpr int degree = block_size * elems_per_thread;
+  for (int z = 0; z < (glwe_dimension + 1); z++) {
```
Same here again.

agnesLeroy commented on concrete-cuda/cuda/src/polynomial/functions.cuh:
```diff
@@ -177,79 +189,88 @@ __device__ void add_to_torus(double2 *m_values, Torus *result) {
-  // Reverse the accumulator
-  // accumulator.reverse_inplace();
+    // Reverse the accumulator
+    // accumulator.reverse_inplace();
```
We could remove those commented lines maybe?

pdroalves replied:
These comments were quite useful to me to understand what is being done in each part of the function. I would keep them.

agnesLeroy replied:
I was confused about line 219 which is pseudo code, but we don't actually have a reverse_inplace() function in the code base. Maybe replacing the pseudo code with an explanation would be clearer?

agnesLeroy commented on concrete-cuda/cuda/src/polynomial/functions.cuh:
```diff
@@ -177,79 +189,88 @@ __device__ void add_to_torus(double2 *m_values, Torus *result) {
-  // Perform ACC * X
-  // accumulator.multiply_by_monomial_negacyclic_inplace(1);
+    // Perform ACC * X
+    // accumulator.multiply_by_monomial_negacyclic_inplace(1);
```
Maybe we could remove these commented lines as well?

agnesLeroy commented on concrete-cuda/cuda/src/polynomial/functions.cuh:
```diff
@@ -177,79 +189,88 @@ __device__ void add_to_torus(double2 *m_values, Torus *result) {
-  // Copy to the mask of the LWE sample
-  // accumulator.copy_into(lwe_array_out);
+    // Copy to the mask of the LWE sample
+    // accumulator.copy_into(lwe_array_out);
```
Here again line 266 could be removed.

agnesLeroy reviewed (changes requested):
Hi @pdroalves! Thanks a lot for this PR, the code looks much more readable to me now! I have a few remarks you'll see, nothing major :slightly_smiling_face: You did an amazing job!

pdroalves commented:
Should we simply ignore the CPU test? I think this is one of those problematic statistical tests.

agnesLeroy commented on concrete-core-fixture/src/fixture/lwe_ciphertext_vector_discarding_bootstrap_3.rs:
```diff
@@ -0,0 +1,335 @@
+        for (x, y) in means.into_iter().zip(actual.into_iter()) {
+            for (expected, obtained) in x.into_iter().zip(y) {
+                if (<Precision::Raw as CastInto<i64>>::cast_into(expected)
+                    - <Precision::Raw as CastInto<i64>>::cast_into(obtained))
```
Is this not the same check as in the fixture 2? Why did you add the fixture 3?

pdroalves replied:
Fixture3 tests the PBS for different parameters, with k>1. I couldn't find a way to check which engine was being tested, so I couldn't avoid using those new parameters on the low latency kernel, which does not support k>1 yet.

agnesLeroy replied:
Oh ok makes sense. It's a shame to add a new fixture then but we can remove it when low lat PBS supports k > 1 as well. No problem for me :slightly_smiling_face:

pdroalves replied:
Totally! It is just temporary.

pdroalves commented:
I added a comment explaining how the `GadgetMatrix` works. Moreover, I also changed all the functions in `functions.cuh` to receive a number of concatenate polynomials instead of `glwe_dimension`, except for the `sample_*` methods since it makes sense to talk about GLWEs in that context.

pdroalves commented:
@slab-ci gpu_test
